pkg/autodiscovery/configmgr.go

```diff
@@ -72,3 +77,353 @@ type configManager interface {
+	// keyed by its serviceID and with its AD identifiers stored separately.
+	// This is the "base truth" of services -- the set of new services
+	// processed net deleted services.
+	activeServices map[string]serviceAndADIDs
+
+	// templatesByADID catalogs digests for all templates, indexed by their AD
+	// identifiers.  It is an index to activeConfigs.
+	templatesByADID multimap
+
+	// servicesByADID catalogs serviceIDs for all services, indexed by their AD
+	// identifiers.  It is an index to activeServices.
+	servicesByADID multimap
+
+	// serviceResolutions maps a serviceID to the resolutions performed for
+	// that service: serviceID -> template digest -> resolved config digest.
+	serviceResolutions map[string]map[string]string
+
+	// scheduledConfigs contains an entry for each scheduled config, keyed by
+	// its digest.  This is a mix of resolved templates and non-template
+	// configs.  The returned configChanges from interface methods correspond
+	// exactly to changes in this map.
+	scheduledConfigs map[string]integration.Config
+}
+
+var _ configManager = &reconcilingConfigManager{}
+
+// newReconcilingConfigManager creates a new, empty reconcilingConfigManager.
+func newReconcilingConfigManager() configManager {
+	return &reconcilingConfigManager{
+		activeConfigs:      map[string]integration.Config{},
+		activeServices:     map[string]serviceAndADIDs{},
+		templatesByADID:    newMultimap(),
+		servicesByADID:     newMultimap(),
+		serviceResolutions: map[string]map[string]string{},
+		scheduledConfigs:   map[string]integration.Config{},
+	}
+}
+
+// processNewService implements configManager#processNewService.
+func (cm *reconcilingConfigManager) processNewService(adIdentifiers []string, svc listeners.Service) configChanges {
+	cm.m.Lock()
+	defer cm.m.Unlock()
+
+	svcID := svc.GetServiceID()
+	if _, found := cm.activeServices[svcID]; found {
+		log.Debugf("Service %s is already tracked by autodiscovery", svcID)
+		return configChanges{}
+	}
+
+	//  1. update activeConfigs / activeServices
+	cm.activeServices[svcID] = serviceAndADIDs{
+		svc:   svc,
+		adIDs: adIdentifiers,
+	}
+
+	//  2. update templatesByADID / servicesByADID to match
+	for _, adID := range adIdentifiers {
+		cm.servicesByADID.insert(adID, svcID)
+	}
+
+	//  3. update serviceResolutions, generating changes
+	changes := cm.reconcileService(svcID)
+
+	//  4. update scheduledConfigs
+	return cm.applyChanges(changes)
+}
+
+// processDelService implements configManager#processDelService.
+func (cm *reconcilingConfigManager) processDelService(svc listeners.Service) configChanges {
+	cm.m.Lock()
+	defer cm.m.Unlock()
+
+	svcID := svc.GetServiceID()
+	svcAndADIDs, found := cm.activeServices[svcID]
+	if !found {
+		log.Debugf("Service %s is not tracked by autodiscovery", svcID)
+		return configChanges{}
+	}
+
+	//  1. update activeConfigs / activeServices
+	delete(cm.activeServices, svcID)
+
+	//  2. update templatesByADID / servicesByADID to match
```

Comment:
`templatesByADID` is not updated here. should this be removed from the comment?

Reply:
This is referring to the steps above.  The `/` indicates "or".  I'll update it, and link to those steps.